File: apps/web/src/types/stock.ts
export interface StockSearchResult {
  ticker: string;
  name: string;
}

export interface StockQuote {
  ticker: string;
  price: number | null;
  currency: string;
}

export interface StockInfo {
  ticker: string;
  name: string;
  sector: string;
  industry: string;
  market_cap: string;
  pe_ratio: number | string;
  forward_pe: number | string;
  dividend_yield: number | string;
  beta: number | string;
  week_52_high: number | string;
  week_52_low: number | string;
  avg_volume: number | string;
  volume: number | string;
  employees: number | string;
  website: string;
  summary: string;
  top_institutional: Record<string, unknown>[] | null;
  latest_recommendation: {
    firm: string;
    rating: string;
    action: string;
  } | null;
}

export interface StockFundamentals {
  ticker: string;
  price: number | null;
  sector: string | null;
  market_cap: string | null;
  pe_ratio: string | null;
  beta: number | null;
  dividend_yield: number | null;
  next_earnings_date: string | null;
}

export interface OHLCVBar {
  timestamp: string;
  open: number;
  high: number;
  low: number;
  close: number;
  volume: number;
}

export interface TechnicalIndicators {
  ticker: string;
  current_price: number;
  ma_20: number | null;
  ma_50: number | null;
  ma_200: number | null;
  rsi_14: number | null;
  rsi_signal: string;
  current_volume: number;
  avg_volume: number;
  volume_ratio: number;
  support: number | null;
  resistance: number | null;
}

File: apps/api/app/schemas/stock.py
from datetime import datetime

from pydantic import BaseModel


class StockSearchResult(BaseModel):
    ticker: str
    name: str


class StockQuote(BaseModel):
    ticker: str
    price: float | None = None
    currency: str = "USD"


class StockFundamentals(BaseModel):
    ticker: str
    price: float | None = None
    sector: str | None = None
    market_cap: str | None = None
    pe_ratio: str | None = None
    beta: float | None = None
    dividend_yield: float | None = None
    next_earnings_date: str | None = None


class StockInfo(BaseModel):
    """Comprehensive stock details from yfinance."""

    ticker: str
    name: str = ""
    sector: str = "N/A"
    industry: str = "N/A"
    market_cap: str = ""
    pe_ratio: float | str = "N/A"
    forward_pe: float | str = "N/A"
    dividend_yield: float | str = "N/A"
    beta: float | str = "N/A"
    week_52_high: float | str = "N/A"
    week_52_low: float | str = "N/A"
    avg_volume: int | str = "N/A"
    volume: int | str = "N/A"
    employees: int | str = "N/A"
    website: str = "N/A"
    summary: str = ""
    top_institutional: list[dict] | None = None
    latest_recommendation: dict | None = None


class OHLCVBar(BaseModel):
    timestamp: datetime
    open: float
    high: float
    low: float
    close: float
    volume: int


class TechnicalIndicators(BaseModel):
    ticker: str
    current_price: float
    ma_20: float | None = None
    ma_50: float | None = None
    ma_200: float | None = None
    rsi_14: float | None = None
    rsi_signal: str = "Neutral"  # Overbought / Oversold / Neutral
    current_volume: float = 0
    avg_volume: float = 0
    volume_ratio: float = 1.0
    support: float | None = None
    resistance: float | None = None


class NewsArticle(BaseModel):
    title: str
    url: str
    source: str
    published_at: str
    summary: str
    banner_image: str = ""
    overall_sentiment_score: float | None = 0.0
    overall_sentiment_label: str = ""
    ticker_sentiment_score: float | None = 0.0
    ticker_sentiment_label: str = "Neutral"
    ticker_relevance: float | None = 0.0


class ForecastHistorical(BaseModel):
    dates: list[str]
    prices: list[float]


class ForecastPrediction(BaseModel):
    dates: list[str]
    prices: list[float]
    upper_bound: list[float]
    lower_bound: list[float]


class ForecastModelInfo(BaseModel):
    method: str
    training_period: str
    slope_per_day: float
    r_squared: float


class StockForecast(BaseModel):
    ticker: str
    current_price: float
    forecast_days: int
    trend_signal: str
    predicted_price: float
    price_change: float
    pct_change: float
    rsi: float
    ma_20: float
    ma_50: float
    historical: ForecastHistorical
    forecast: ForecastPrediction
    model_info: ForecastModelInfo
